### PR TITLE
Recursively expand folders when moving

### LINKED ISSUE #105691: Don't delete files when replacing dir
Suppose you have two dirs.

<project>/server/http/middleware

has file test.go

Now you have the 2nd dir.

<project_second>/server/http/middleware

has file test2.go

If you drag the dir /server/http/middleware from <project_second> into <project>, test.go will be deleted.

This is not what I would expect after 15 years of using Windows and Linux. I know the message is clear on it, but UI should be better than that.

Please don't delete the already existing file OR if you do, please make the file recoverable. Just had this issue and had some progress wiped.

### PR DESCRIPTION
**Pull request by plainerman**
This PR will fix #105691, but still requires some discussion with @isidorn 

I have created a somehow working prototype and would like your feedback on the base construct and discuss a few points with you @isidorn.

1. When a file and a folder have the same name, they cannot be merged. What should happen in this case? With my current suggested implementation, the user will receive the classical dialog stating that xxx exists in source and destination and if files should be replaced. I think this is a valid behavior, but what do you think?
2. The `bulkEditService` somehow cannot properly revert changes, if the remaining folders are deleted like I currently do (with `undefined` as target). This causes issues, when files are present in both folders (i.e. in the `catch` block where it is tried again with overwrite).

I was thinking about:

i) Deleting the root folder after the process has been completed. But this can throw an error if the directory is not empty (e.g. if there is another empty directory inside)

ii) First execute all normal moves and then the remaining deletes. But with my testing, the `bulkEditService` does not retain the order of edits causing the same error.

What do you think is the best way to clean up afterwards? Calling bulk edit with a single item each? Or is there a way to delete the folder even if it is not empty?

3. I am open to suggestions for renaming `explodeConflictingFolderEdit`. Naming things is hard 😢

### CODE REVIEW

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts [hunk 1]
```diff
@@ -1499,6 +1510,37 @@ export class FileDragAndDrop implements ITreeDragAndDrop<ExplorerItem> {
...
+		const edits: ResourceFileEdit[] = [];
+
+		const newEdits = resolvedOldResource.children?.map(child => new ResourceFileEdit(child.resource, URI.joinPath(edit.newResource!, child.name)));
+
+		for (const newEdit of newEdits!) {
+			edits.push(...await this.explodeConflictingFolderEdit(newEdit));
+		}
+
+		// Since all children of the folder are now moved individually, we have to delete the folder
+		edits.push(new ResourceFileEdit(edit.oldResource, undefined));
```
Comment (plainerman) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:1539]: In here, the deletion will be recursively added.
But as I said in the text, this causes issues.

**Comment by isidorn** [comment]
This week we are in our endgame process ,which means we are wrapping up our January milestone and we are not accepting new changes. So I plan to review this in more detail next week. 
For now I will only quickly comment on some of your questions:
1) A dialog in that case makes sense. The dialog should give the option to: "Replace", "Always Replace", "Keep Original", "Always Keep Original" - or something like that
2) this is a known limitation of the bulk edit service, don't worry about if for this PR, we have other issue tracking this. Ideally you would call the `bulkEditService` the least number of times possible.
3) `explode` name works fine for now, we can polish this later

**Comment by plainerman** [comment]
Thanks for the quick response, and don't worry about the endgame.
I have resolved deleting the orphaned folders and made some polishes. But let's discuss the code itself next week.

Anyway, I would still have a followup question regarding 1)
When we have conflicting files, we list all conflicts. This behavior has been unchanged and is currently also triggered for my changes.

https://github.com/microsoft/vscode/blob/7b0cfdd04ad530a9c8d782c618afd79290be3d64/src/vs/workbench/contrib/files/browser/views/explorerViewer.ts#L1475-L1484

image

You suggested to implement this on a per-file-basis (Replace, Replace All, Skip, Skip all).

Windows behaves in the following way:
image

I think we could do something similar as in Windows.

**Now my question**:
Should I implement this feature only when folders are merged (i.e. my code is triggered), or for all conflicts in general?
I think implementing for all moves would be beneficial to users (and easier for me to implement).

And if we decide on implementing this for all moves, should we do this in a new issue / PR (which I would be happy to create and implement based on this branch) or just continue in here.

Apart from that, the only thing left for me to do is to make a setting, whether folders should be merged or replaced.

Thank you for your input!

**Comment by isidorn** [comment]
1) conflicts - I suggest that we use what we have already. That way we are consistent and no need to write new code
2) for all merges is great if it is easier, and we would be consistent
3) just continue here, no need for a new PR. Let's try to keep it compact
4) here's a potential name of the setting `explorer.onFolderConflict`: `merge` | `replace`. Just an idea...

**Comment by isidorn** [comment]
Also this week I will have time so let me know when this is in a decent shape so I test it out and check out the code.
Thanks a lot for your help!

**Comment by plainerman** [comment]
Thank you very much for your reply @isidorn. I have just implemented your suggested naming for the property and I am about to implement the new replacement dialog.

As suggested above, I would create a popup similar to Windows (Replace, Skip, Decide).

Currently, we use the `IDialogService` to create the popup.

https://github.com/microsoft/vscode/blob/3088c27fbdcbf2318385d4a4be5f3a0f02e994a6/src/vs/platform/dialogs/common/dialogs.ts#L200

Unfortunately, as I see it, `IConfirmation` only allows us to place 2 buttons, whereas we would need 3 (+1 cancel).

So i thought about using `show` instead of `confirm`

https://github.com/microsoft/vscode/blob/3088c27fbdcbf2318385d4a4be5f3a0f02e994a6/src/vs/platform/dialogs/common/dialogs.ts#L249

But I believe `show` does not allow me to add a detail text (like listing the files above).

Could you point me into the right direction on how to create such a dialog with 3+1 buttons and a details pane?
Or should we leave out listing all files (like Windows does it)?

**Comment by isidorn** [comment]
@plainerman yeah you should use what the `dialogService` gives you.
`show` has the `options` object which has the `detail`. So hopefully this should just work
Code pointer https://github.com/microsoft/vscode/blob/7c90e1f5ba949d201d333306d60a980380760326/src/vs/platform/dialogs/common/dialogs.ts#L249

**Comment by plainerman** [comment]
Hey @isidorn,
From my point of view, the code is now ready for a review and I am open to any discussions.

You can use this minimal [example.zip](https://github.com/microsoft/vscode/files/5924268/example.zip) I have used in my tests, where you move the folder `middleware` into the folder `server`. 

After we have agreed on the code, I can also add some unit / integrations tests if you'd like. But for that I'd need some code pointers on how the test should look like and where to add.

Thank you :)

**Comment by isidorn** [comment]
@plainerman thanks a lot for working on this. I like the changes and I think the flow makes sense.
What I do not like is the fact that this introduces a lot of new code, I understand that it might not have been possible to do it any other way, but just looking at it these are 200 new lines for this "simple" feature.
I understand that it might not sound like a lot, but maintaining this code might be the problem going forward for me since I was not the author.

Bottom line: I like the work, but if we could simplify it somehow would make my decision to merge much easier.
This way I plan to review it in more details next week or the week after.

Hope this makes sense and sorry if I put on you on the wrong path.

**Comment by plainerman** [comment]
@isidorn thank you for your honest reply :)

About 60 of the newly added lines (more than a third) are for creating the individual dialogs. The actual change for merging directories and adding a setting for this was a lot smaller (+55/-25 or something similar).

Anyhow, here is my suggestion on how we continue:
* I will see what I can do to make the code more concise without impacting readability. I will also add more comments, where needed so that you do not have to worry about maintaining the code.
* After that I will ping you and if you still think that this PR / feature is too large, we could split it. I revert the changes here that involve the interactive user handling and create a new PR with a corresponding issue for it. The current PR would only contain the new setting and would still feature the old "replace" dialog with the interactive code in the new issue.
* If it is still too complex but prefer keeping everything in this PR, I could elaborate more on the thought process and describing the idea. This should help you get more familiar with the code.

**Comment by isidorn** [comment]
@plainerman sounds good. Making it more compact sounds like a good idea if possible.
You do not have to over-do on the comments.

Please ping me again and I will review in deeper detail.

**Comment by plainerman** [comment]
@isidorn I could make the code in some areas more compact (employing different approaches, simplifying UI code, etc.) to the point where I believe that further minimizing it could negatively impact readability. 

What is your take with the current complexity? Should we split the PR? And as this is now simpler, will you be able to look at it sooner?

Keep in mind, that creating the UI dialogs requires about 35 added lines of code.

Thank you very much :)

File: src/vs/workbench/contrib/files/browser/files.contribution.ts [hunk 2]
```diff
@@ -476,6 +476,15 @@ configurationRegistry.registerConfiguration({
 			'description': nls.localize('compressSingleChildFolders', "Controls whether the explorer should render folders in a compact form. In such a form, single child folders will be compressed in a combined tree element. Useful for Java package structures, for example."),
 			'default': true
 		},
+		'explorer.onFolderConflict': {
+			enum: ['merge', 'replace'],
+			enumDescriptions: [
+				nls.localize('merge', "The conflict will be resolved recursively, meaning that all children will be moved individually. This is the default behavior in most operating systems"),
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/files.contribution.ts:482]: "This is the default behavior" not needed sentence, especially since this behavior is not like this on macOS. So please remove this sentence.

First sentence needs to be clarified. For example: "files from both folders will be merged into the new destination folder" or something like that

File: src/vs/workbench/contrib/files/browser/views/explorerViewer.ts [hunk 3]
```diff
@@ -748,17 +748,39 @@ function getFileOverwriteConfirm(name: string): IConfirmation {
...
 
-	return getFileOverwriteConfirm(basename(files[0]));
+function getMultipleFilesOverwriteOrSkip(files: URI[]): { severity: Severity, message: string, buttons: string[], options: IDialogOptions } {
+	return {
+		severity: Severity.Warning,
+		message: localize('confirmManyOverwrites', "The following {0} files and/or folders already exist in the destination folder?", files.length),
+		buttons: [
+			localize({ key: 'replaceAllButtonLabel', comment: ['&& denotes a mnemonic'] }, "&&Replace all"),
+			localize({ key: 'skipAllButtonLabel', comment: ['&& denotes a mnemonic'] }, "&&Skip all"),
+			localize({ key: 'decideAllButtonLabel', comment: ['&& denotes a mnemonic'] }, "&&Decide invidivually"),
```
Comment (isidorn) [inline src/vs/workbench/contrib/files/browser/views/explorerViewer.ts:776]: Let's skip this option "Decide individually". Makes the whole thing overly comples and I do not like the idea of showing a lot of dialogs. Windows explorer does this by showin a dialog with a bunch of checkmarks - which we do not want to do right now.

**Review by isidorn — CHANGES_REQUESTED** [review]

**Comment by isidorn** [comment]
Did a first quick review of the PR.
It would be great if this also worked when Drag and Drop from the desktop. Since currently it only works when drag and dorp from inside the Explorer. And ideally those two experiences would be the same.

I have also found out that the mac finder behaves like the explorer currently does, before this PR. So leaving it like it is is not the worst solution.

Apart from that: we do not ask the dialog for deeply rooted children correct, we just check for differences on the first level? Which I think we should just do it on the first level.

File: src/vs/workbench/contrib/files/browser/files.contribution.ts (hunk 2 above)
Comment (plainerman) [reply src/vs/workbench/contrib/files/browser/files.contribution.ts]: I have changed the text, hopefully it is better now.

**Comment by plainerman** [comment]
Thanks @isidorn for giving me first suggestions on my changes! :)
I have resolved your requested changes. 

I will look into adding support for desktop drag-and-drops. Couple of questions:
1) Should `handleWebExternalDrop` also make use of this functionality? If so, how can I simulate and test this event?
2) In editor drops: we move the files, `catch` FIle-Move-Conflicts, ask the user, and maybe try again. In drops from the OS, we check for conflicting files *before moving* (and also ask the user). Is there any reason why we don't check for conflicts before moving in both cases?

If I can check for errors beforehand in both cases, my life (and probably the code too) would be easier, as the behavior would be nearly identical in both cases.

Regarding your question about deeply rooted children, I am not quite sure if I understand you correctly.
Assuming this example, where we move `middleware` into `other` (with merge folders activated):
```
middleware
  > file1
  | subfolder
    > file2
    > file3
other
  | middleware
    > file1
    | subfolder
      > file2
      > file4
```

It will show, that there is a conflict for file1 and file2. With merge enabled, it will never show a folder as a conflict, as those will be merged.
I believe, in general, we should list all conflicts (even in deeper areas), as those could be essential files that are overriden.

Only showing e.g. `subfolder` as a conflict would be confusing as file3 and file4 do not cause a conflict. This replacement could then be misinterpreted as "overriding" the folder.

Looking forward to your reply, so I can continue working on this issue.

**Comment by plainerman** [comment]
Hey @isidorn,

I have just added support for drops from the explorer. Regarding question 2) I have adapted the current behavior from editor-drops. Meaning that we use a try then fix approach as used here (editor drops)

https://github.com/microsoft/vscode/blob/d29d7e023b6085a3cf73a3a04bc18b62a6ea5fef/src/vs/workbench/contrib/files/browser/views/explorerViewer.ts#L1472-L1480

instead of check then execute approach (explorer drops)

https://github.com/microsoft/vscode/blob/d29d7e023b6085a3cf73a3a04bc18b62a6ea5fef/src/vs/workbench/contrib/files/browser/views/explorerViewer.ts#L1336-L1344

But this can be changed rather easily, if you'd like the behavior to be switched. Making it one way and the other would make the code a bit more complex / add more lines.

I am still looking forward to discussing the remaining points with you, otherwise the PR is ready to go again.

**Comment by isidorn** [comment]
Thanks for the updates, assigning to March to review then

**Comment by plainerman** [comment]
@isidorn I would love to continue discussion of this feature, my code, and hopefully get it merged soon. Is the April milestone still holding?

Thank you

**Comment by isidorn** [comment]
@plainerman thanks a lot for your ping, and sorry for the slow response. I plan to merge this in next week, so this will happen beginning of May so we have a full milestone to self host on it and see how it works

**Comment by plainerman** [comment]
Great, thank you very much.
Just ping me if there is anything I can help you with!

**Comment by plainerman** [comment]
@isidorn I saw that you have pushed it back, any updates?

**Comment by isidorn** [comment]
@plainerman yes, I tried it out and some thing were not working as I expected.
Original folder would get left behind empty when there is a merge. Also the native Mac Finder does not offer the option to merge, and this was not that highly requested feature by the users. So I lowered the priority of this PR, since I have other things to look into May. Sorry about that!

**Comment by plainerman** [comment]
@isidorn I can take a look into this and see if I can reproduce and fix it.
If you still have the folder hierarchy that caused the problem, reproducing it would be easier.

**Comment by isidorn** [comment]
As I mentioned in the issue: "We are transferring the ownership of a file explorer to a new engineer. So I am cleaning up issues which I am not 100% certain that we want to tackle. Due to that I am closing this as out of scope, so we have a clean slate.
We might reopen this in the future if we receive more feedback."

I am sorry that we did not get to merge in this pr, and I apologize if my communication was not very clear from the start.
